Sources/Core/AWSClientRuntime/Auth/CredentialsProviders/ECSCredentialsProvider.swift

```diff
@@ -0,0 +1,96 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import AwsCommonRuntimeKit
+import ClientRuntime
+import Foundation
+
+/// A credentials provider that sources credentials from ECS container metadata
+public struct ECSCredentialsProvider: CredentialsSourcedByCRT {
+    let crtCredentialsProvider: CRTCredentialsProvider
+    /// Creates a credential provider that sources credentials from ECS container metadata
+    /// ECS creds provider can be used to access creds via either relative uri to a fixed endpoint http://169.254.170.2,
+    /// or via a full uri specified by environment variables:
+    /// - AWS_CONTAINER_CREDENTIALS_RELATIVE_URI
+    /// - AWS_CONTAINER_CREDENTIALS_FULL_URI
+    /// - AWS_CONTAINER_AUTHORIZATION_TOKEN
+    ///
+    /// If both relative uri and absolute uri are set, relative uri has higher priority.
+    /// Token is used in auth header but only for absolute uri.
+    /// - Returns: `CredentialsProvider`
+    /// - Throws: CommonRuntimeError.crtError or InitializationError.missingURIs
+    public init(
+        environment: Environment = ProcessEnvironment()
+    ) throws {
+        let relativeURI = environment.environmentVariable(key: "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
+        let absoluteURI = environment.environmentVariable(key: "AWS_CONTAINER_CREDENTIALS_FULL_URI")
+
+        guard relativeURI != nil || isValidAbsoluteURI(absoluteURI) else {
```

Comment:
I think this is a bit confusing because to determine when this guard block will be executed, we really gotta think through the conditional because guard executes for false condition, and we're using `||` operator with `!=` on the relativeURI.

Maybe more direct way is to do: 
`if (relativeURI == nil && !isValidAbsoluteURI(absoluteURI)) { \\ throw error }` ?
In this case it's much simpler (imo) to think it through - we only throw error if `relativeURI` is nil & absoluteURI is invalid.

But in general I'm still in process of trying to be friends with guard syntax - so I'm not entirely sure how much of me finding this confusing is due to the difference in my familiarity with if vs. guard. Also, using guard to check for exception case seems more idiomatic approach in Swift since we're literally "guarding" it haha. Let me know what you think.

Reply:
This is a great comment and in general I agree that yours is more readable at the moment (to both you and me), but I have a strong preference for `guard` here. `if` statements are a general-purpose conditional, so anything could be inside of such a code block. If I were only setting state here I would switch it, but `guard` indicates that this is a required condition for the program to continue. So by using guard I am emphasizing that we cannot continue with execution unless we have a valid relative or absolute URI.

Reply:
This can be read as "To continue executing (guard) we require either relativeURI is not nil (!=) or (||) absoluteURI is valid, otherwise we'll throw an error"

Reply:
> "To continue executing (guard) we require either relativeURI is not nil (!=) or (||) absoluteURI is valid, otherwise we'll throw an error"

Reading it that way does make it much better, and yep, I agree that using `guard` here does put better emphasis on execution condition.

Reply:
Usual purpose of guard is to verify initial conditions for the following block of code.  This looks appropriate.